helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java

```diff
@@ -0,0 +1,132 @@
+    //           /\
+    //          /  \
+    //         /    \
+    //       JOB1   JOB2
+
+    JobConfig.Builder jobBuilder = JobConfig.Builder.fromMap(WorkflowGenerator.DEFAULT_JOB_CONFIG)
+        .setMaxAttemptsPerTask(1).setWorkflow(workflowName)
+        .setJobCommandConfigMap(ImmutableMap.of(MockTask.JOB_DELAY, "1000"));
+
+    JobConfig.Builder jobBuilder2 = JobConfig.Builder.fromMap(WorkflowGenerator.DEFAULT_JOB_CONFIG)
+        .setMaxAttemptsPerTask(1).setWorkflow(workflowName)
+        .setJobCommandConfigMap(ImmutableMap.of(MockTask.JOB_DELAY, "1000"));
+
+    JobConfig.Builder jobBuilder3 = JobConfig.Builder.fromMap(WorkflowGenerator.DEFAULT_JOB_CONFIG)
+        .setMaxAttemptsPerTask(1).setWorkflow(workflowName)
+        .setJobCommandConfigMap(ImmutableMap.of(MockTask.JOB_DELAY, "1000"));
+
+    builder.addParentChildDependency("JOB0", "JOB1");
+    builder.addParentChildDependency("JOB0", "JOB2");
+    builder.addJob("JOB0", jobBuilder.setExecutionDelay(executionDelay));
+    builder.addJob("JOB1", jobBuilder2);
+    builder.addJob("JOB2", jobBuilder3);
+
+    // Since the execution delay is set as a long time, it is expected to reach time out.
+    // If the workflow completed without any exception, it means that ExecutionDelay has not been
+    // respected.
+    _driver.start(builder.build());
+
+    // Verify workflow Context is created.
+    // Wait until Workflow Context is created.
+    boolean workflowCreated = TestHelper.verify(() -> {
+      WorkflowContext wCtx1 = _driver.getWorkflowContext(workflowName);
+      return wCtx1 != null;
+    }, 30 * 1000);
+    Assert.assertTrue(workflowCreated);
+
+    // Verify StartTime is set for the Job0.
+    // Wait until startTime is set.
+    boolean startTimeSet = TestHelper.verify(() -> {
+      WorkflowContext wCtx1 = _driver.getWorkflowContext(workflowName);
+      long startTime = -1;
+      try {
+        startTime = wCtx1.getJobStartTime(TaskUtil.getNamespacedJobName(workflowName, "JOB0"));
+      } catch (Exception e) {
+        // pass
+      }
+      return startTime != -1;
+    }, 30 * 1000);
+    Assert.assertTrue(startTimeSet);
+
+    // Verify JobContext is not created. Creation of the JobContext means that the job is ran or
+    // in-progress. The absence of JobContext implies that the job did not run.
+    boolean isJobStarted = TestHelper.verify(() -> {
+      JobContext jobCtx0 =
+          _driver.getJobContext(TaskUtil.getNamespacedJobName(workflowName, "JOB0"));
+      JobContext jobCtx1 =
+          _driver.getJobContext(TaskUtil.getNamespacedJobName(workflowName, "JOB1"));
+      JobContext jobCtx2 =
+          _driver.getJobContext(TaskUtil.getNamespacedJobName(workflowName, "JOB2"));
+      return (jobCtx0 != null || jobCtx1 != null || jobCtx2 != null);
+    }, 30 * 1000);
+    Assert.assertFalse(isJobStarted);
+
+
+    Assert.assertNull(_driver.getWorkflowContext(workflowName).getJobState(TaskUtil.getNamespacedJobName(workflowName, "JOB0")));
```

Comment:
We seem to be making a ZK read call for each check. Is it possible to just do one read and do the checks on one read?